### DISCUSSION TITLE
Coding4Engineers Book feedback

### DISCUSSION THREAD

**Discussion by Mech-eng-au** [leap71/PicoGK#42] [Coding for (Computational) Engineers]
Hi Lin,

The book is really coming together! Every time I read a chapter, my understanding of both OOP (object-oriented programming) and C# improves, and I become more proficient with PicoGK.

I thought it would be helpful to open this discussion so that the computational engineering community can provide feedback and help you proofread the Coding4Engineers book.

Here's an example from your latest chapter (11):

When writing the constructor for the BasePlate class, you use `Vector` instead of `Vector2`. This will result in a `CS1729` error when Fixture.BasePlate is called (see screenshot below).

![image](https://github.com/user-attachments/assets/c4b8b60e-207e-4729-b030-00ee199b803d)

As I'm new to programming, I apologize if my terminology isn't accurate. I hope that it makes sense anyway.

**Comment by LinKayser** [comment]
Thank you — I fixed the mistake in the book example!
Lin

**Comment by ahuller1991** [comment]
Thank you for creating and making open-source both PicoGK and this book. I sincerely agree with your philosophy on current CAD software vs. physics-driven design. I have made it through the chapters you have posted thus far, and I have been able to create a rocket nozzle using a combination of NASA CEA, Compressible flow, and Rao Contour equations. I still have a lot to learn (I am not proficient in C#) but this has been a lot of fun, and I can see the power in this capability. I am looking forward to future chapters of your book!
<img width="422" alt="RocketNozzle" src="https://github.com/user-attachments/assets/81adfa7e-0a2c-4d25-95ab-0c088ccecdfe" />

**Reply by TristanTerran** [reply]
Im looking to do the same type of thing for my student rocketry team. could you give me a hint to figure out the logic structure of the code?

**Comment by digitaldoctor** [comment]
Greetings, ALL and wishes for a most productive New Year.

I would appreciate feedback on a line of though that I'm parsing, as it pertains to Computational Engineering and C#. The line is driven by an earnest interest in democratizing awareness of, and access to, the likes of CE.

1) in the age of AI, is it absolutely necessary to go through the gauntlet of learning C# from the ground up?

2) is there a sound skeletal understanding of CE principles that can be had, and then the hardcore coding be offloaded to a C# coding assistant?

3) if the answer to question 2 is yes, then what does that fundamental knowledge base look like? Can it be presently so clearly and simply that high school and community college students master it?

Looking forward to constructive responses.

Rafiki Cai

**Reply by digitaldoctor** [reply]
It may have been better, if I had read the Preamble to the CFE book. LOL

     "you should **not** attempt to encode logic into an LLM". (Lin Kayser)

This seems to be exactly what billions of dollars are being deployed to do. There can be no AGI
without logic being wired into LLMs.

If the basic workload of logic is to process information systematically, draw valid conclusions, and avoid contradictions,
why shouldn't this capacity exist in LLMs.

**Reply by LinKayser** [reply]
Logic should be encoded in computational models, they are computationally light, and always give you the right answer. You can bombard an LLM with many petabytes of question-answer pairs, and it may still give you a wrong answer at times. Moreover, you cannot be *sure* that it will give you the right answer. LLMs are great for things that are not easily explainable through logic (natural language, "good" driving, why one person in a picture is a man, and another a woman).

The people spending the billions largely know that and are looking at other ways to build general intelligence. ChatGPT, for example, now synthesizes answers as code, and then runs the code (how many days to Christmas?), which is an interesting way to do it, but probably not the final way. If you read some of the public writings of Yann LeCun, you can follow these arguments.

Stephen Wolfram also has some good takes on it. I have summarized some of it here: 

https://jlk.ae/2024/01/01/on-tribal-knowledge-llms-and-computational-knowledge-in-engineering/

Best,
Lin

**Comment by digitaldoctor** [comment]
TYPO In **FUNDAMENTALS** (Second Paragraph)
------------------
" that we will **built** upon in the following chapters""

**Comment by eherrmann8** [comment]
The book is definitely helpful. I am looking for a comprehensive list of functions (and examples) within PicoGK & ShapeKernal so that I don't have to surf through the code to figure out what is available outside of what you have already put in the chapter examples. Also I'm looking for functions like volume, volumetric center and moment of inertia for a given closed voxel object.

**Reply by eherrmann8** [reply]
I found the CalculateProperties function which returns the volume of the voxel field. Can this be extended to also return the volumetric center and moments of inertia about the local frame coordinates?

**Comment by dvd101x** [comment]
Hi, thanks for the book.

I found a typo in [Chapter 4: Classes: Class Definition](https://picogk.org/coding-for-engineers/4-classes.html#class-definition) last sentence.

> Now, it’s time to instatiate our first object!

Should be:
> Now, it’s time to **instantiate** our first object!

Missing an "n"

**Reply by dvd101x** [reply]
Hi, just a few more typos

| phrase | fix | to | source |
| --- |---|---|---|
| it **should’t** work | should’t | shouldn’t | [Preamble: My Story](https://picogk.org/coding-for-engineers/#preamble-my-story)
| I found it **extremly** hard | extremly | extremely | [Fundamentals - The skeleton of a class](https://picogk.org/coding-for-engineers/2-fundamentals.html#the-skeleton-of-a-class)
| one very **straighforward** one | straighforward | straightforward | [Fundamentals - Constructors](https://picogk.org/coding-for-engineers/2-fundamentals.html#constructors)
| Let’s do the **straighforward** one first  | straighforward | straightforward | [Fundamentals - Constructors](https://picogk.org/coding-for-engineers/2-fundamentals.html#constructors)
| to expose the **mimimal** interface | mimimal | minimal | [Fundamentals - Visibility](https://picogk.org/coding-for-engineers/2-fundamentals.html#visibility) |
| compiled for different **archictures** | archictures | architectures | [Running code](https://picogk.org/coding-for-engineers/3-running-code.html)
| And basically **everyhing** in our current | everyhing | everything | [Inheritance - Summary](https://picogk.org/coding-for-engineers/5-inheritance.html#summary)
| to form a class **hiearchy** | hiearchy | hierarchy | [Inheritance - Summary](https://picogk.org/coding-for-engineers/5-inheritance.html#summary)
| the name of the function and **it’s** signature | it’s | its | [Interfaces - Has a…, is a…, behaves like a…](https://picogk.org/coding-for-engineers/6-interfaces,html#has-a-is-a-behaves-like-a)
| **Slighly** confusingly, there is no | Slighly | Slightly | [Interfaces - Has a…, is a…, behaves like a…](https://picogk.org/coding-for-engineers/6-interfaces.html#has-a-is-a-behaves-like-a)
| being a class of the **interited** type| interited | inherited | [Interfaces - Summary](https://picogk.org/coding-for-engineers/6-interfaces.html#summary)
| you can **stil** agree | stil | still | [Design an aircraft in an afternoon - Designing an airplane using interfaces, placeholders, and stubs](https://picogk.org/coding-for-engineers/7-design-an-aircraft-in-an-afternoon.html#designing-an-airplane-using-interfaces-placeholders-and-stubs)
| maybe we **should just used** a bool value | should just used | "should just have used" or "should just use" | [Design an aircraft in an afternoon - Designing an airplane using interfaces, placeholders, and stubs](https://picogk.org/coding-for-engineers/7-design-an-aircraft-in-an-afternoon.html#designing-an-airplane-using-interfaces-placeholders-and-stubs)
| attachment point **and and** mechanical | and and | and | [Design an aircraft in an afternoon - Geometry is the last thing on your mind](https://picogk.org/coding-for-engineers/7-design-an-aircraft-in-an-afternoon.html#geometry-is-the-last-thing-on-your-mind)
| wings were just flat **recangles** | recangles | rectangles | [ Design an aircraft in an afternoon - Building an airplane](https://picogk.org/coding-for-engineers/7-design-an-aircraft-in-an-afternoon.html#building-an-airplane)
| something that is often **suprisingly** difficult to do | suprisingly | surprisingly | [Design an aircraft in an afternoon - Summary](https://picogk.org/coding-for-engineers/7-design-an-aircraft-in-an-afternoon.html#summary)
| the code you are executing is **interupted** immediately| interupted | interrupted | [First steps in PicoGK - Catching exceptions](https://picogk.org/coding-for-engineers/8-first-steps-in-picogk.html#catching-exceptions)
| Building a fixture is not a **trival** task| trival | trivial | [Let’s build a computational fixture maker (Part 1)](https://picogk.org/coding-for-engineers/9-computational-fixture-maker.html)
| make sure it sits at a **specifed** height above | specifed | specified | [Let’s build a computational fixture maker (Part 1) - Loading and positioning a mesh](https://picogk.org/coding-for-engineers/9-computational-fixture-maker.html#loading-and-positioning-a-mesh)
| you should **keept** it alive | keept | kept | [Let’s build a computational fixture maker (Part 1) - Cleaning up our namespace with nested classes](https://picogk.org/coding-for-engineers/11-computational-fixture-maker-3.html#cleaning-up-our-namespace-with-nested-classes)
| But **we we** can make | we we | we | [Let’s build a computational fixture maker (Part 3) - Drilling holes](https://picogk.org/coding-for-engineers/11-computational-fixture-maker-3.html#drilling-holes)
| and then create the **modfied** version | modfied | modified | [Let’s build a computational fixture maker (Part 3) - Drilling holes](https://picogk.org/coding-for-engineers/11-computational-fixture-maker-3.html#drilling-holes)
| We have **alread** worked with all | alread | already | [Computational geometry (Part 1)](https://picogk.org/coding-for-engineers/13-computational-geometry-part1.html)
| Of **coutse**, In a physically correct manifold | coutse | course | [Computational geometry (Part 2)](https://picogk.org/coding-for-engineers/14-computational-geometry-part2.html)
| let’s have the actual **implementaton** | implementaton | implementation | [Computational geometry (Part 4)](https://picogk.org/coding-for-engineers/16-computational-geometry-part4.html)
| | No link for "Modulating a grid" | Add link for "Modulating a grid" | [Computational geometry (Part 4)](https://picogk.org/coding-for-engineers/16-computational-geometry-part4.html)

**Comment by raaed33** [comment]
In this code snippet in chapter 9(making a fixture for a teapot),
<img width="1017" height="457" alt="image" src="https://github.com/user-attachments/assets/7fa1ada5-8eef-4d29-b36c-be7ec49451aa" />
passing the object `oFixtureBounds` to `Utils.mshCreateCube` throws an error because `mshCreateCube` requires 2 vector3s as input as shown in `PicoGK_Utils.cs` here:

<img width="647" height="96" alt="image" src="https://github.com/user-attachments/assets/ebc4b64c-6883-4b28-89ca-6b82e20de3bd" />

A not so elegant fix I came up with is :
```

Vector3 vecmin = oFixtureBounds.vecMin;
Vector3 vecmax = oFixtureBounds.vecMax;
Vector3 vecscale = vecmax - vecmin; 
Vector3 vecoffset = (vecmin + vecmax) * 0.5f;
Mesh mshIntersect = Utils.mshCreateCube(vecscale,vecoffset);
```

This makes sure `mshCreateCube` gets the vector3 arguments properly.

**Reply by LinKayser** [reply]
Should work with the current version of PicoGK 1.7.7.5 and up, which allows passing a `BBox3`.
Lin

**Comment by hugovdkort** [comment]
Right at the start of the CH10 paragraph ["a class for fixtures"](https://picogk.org/coding-for-engineers/10-computational-fixture-maker-2.html#a-class-for-fixtures) there is a typo in the code:

`public voxAsVoxels()
    {
        return m_voxFixture;
    }`
should be
`public Voxels voxAsVoxels()
    {
        return m_voxFixture;
    }`

Afterward it is (return)typed correctly.

**Reply by LinKayser** [reply]
Thanks, fixed it.

**Comment by DavidDonelan** [comment]
Hi,

I liked the approach of the fixture coding: Step by step and tracking back and optimizing along the way. It's practical in the sense that it gives you what you need when you need it, i.e the basics of object orientation and coding style before the geometry chapters and then more details on C# at the end of the book. 

The book motivated me to start using Git and GitHub, even though it's not a topic the book directly covers.

Best,

DD

**Comment by KofiPhD2019** [comment]
I found a typo in the fourth Summary point for Chapter 8 - First steps in PicoGK.

It reads "One way to create geometry is to use lattices. Lattices consist of one or **man** beams and spheres."

I think you meant to type one or **many** beams.

Regards,
